tests/dragonfly/replication_test.py

```diff
@@ -1068,3 +1067,107 @@ async def test_readonly_script(df_local_factory):
+            await c_master.execute_command(f"SET {key} 0")
+            start = time.time()
+            while time.time() - start < 20:
+                try:
+                    value = await c_master.execute_command(f"INCR {key}")
+                except (redis.exceptions.ConnectionError, redis.exceptions.ResponseError) as e:
+                    break
+            else:
+                assert False, "The incrementing loop should be exited with a connection error"
+            return key, value
+
+        async def block_during_takeover():
+            "Add a blocking command during takeover to make sure it doesn't block it."
+            # TODO: We need to make takeover interrupt blocking commands.
+            return
+            try:
+                await c_blocking.execute_command("BLPOP BLOCKING_KEY1 BLOCKING_KEY2 10")
+            except redis.exceptions.ConnectionError:
+                pass
+
+        async def delayed_takeover():
+            await asyncio.sleep(1)
+            await c_replica.execute_command(f"REPLTAKEOVER 5")
+
+        _, _, *results = await asyncio.gather(delayed_takeover(), block_during_takeover(), *[counter(f"key{i}") for i in range(16)])
+        assert await c_replica.execute_command("role") == [b'master', [[b'127.0.0.1', bytes(str(replica3.port), 'ascii'), b'stable_sync']]]
+
+        for key, client_value in results:
+            replicated_value = await c_replica.get(key)
+            assert client_value == int(replicated_value)
+            replicated_value2 = await c3.get(key)
+            assert client_value == int(replicated_value2)
+
+
+@pytest.mark.parametrize("master_threads, replica_threads", take_over_cases)
+@pytest.mark.asyncio
+async def test_take_over_seeder(df_local_factory, df_seeder_factory, master_threads, replica_threads):
+    master = df_local_factory.create(proactor_threads=master_threads,
+                                     port=BASE_PORT,
+                                     #  vmodule="journal_slice=2,dflycmd=2,main_service=1",
+                                     logtostderr=True)
+    replica = df_local_factory.create(
+        port=BASE_PORT+1, proactor_threads=replica_threads)
+    df_local_factory.start_all([master, replica])
+
+    seeder = df_seeder_factory.create(port=master.port, keys=1000, dbcount=5, stop_on_failure=False)
+    async with (
+        master.client() as c_master,
+        replica.client() as c_replica,
+    ):
+        await c_replica.execute_command(f"REPLICAOF localhost {master.port}")
+        await wait_available_async(c_replica)
+
+        async def seed():
+            try:
+                await seeder.run(target_ops=3000)
+            except (redis.exceptions.ConnectionError, ConnectionRefusedError, Exception, SystemExit):
+                pass
+
+        fill_task = asyncio.create_task(seed())
+
+        await c_replica.execute_command(f"REPLTAKEOVER 5")
+
+        assert await c_replica.execute_command("role") == [b'master', []]
```

Comment:
When the master shuts down we save a snapshot, we can in the test check that all the data in the master passed to the replica if we start the master again and compare the master data to replica data with seeder.capture() and seeder.compare

Reply:
Turns out I had a bug with shutdown snapshotting! good catch :)